### PR TITLE
api endpoint for form letters

### PR DESCRIPTION
**Pull request by eileen-nava**
[Link to ZenHub issue: Create an API Endpoint for the Form Letters Send section of Team Management.](https://github.com/usdoj-crt/crt-portal-management/issues/1254)

## What does this change?
This creates an endpoint that shares information about the form letters sent.

## Checklist:
Do the below while logged in as an admin:
[ ] Navigate to the `/admin` page and select a "date range" which has letters sent in it.  Note the section for the report associated with the letters.
[ ] Navigate to the `api/form-letters/?assigned_section={assigned_section}&?start_date={start_date}&end_date={end_date}` endpoint and enter `{assigned_section}`, `{start_date}` and `{end_date}` values that match the values from the admin page's date range.
[ ] Confirm that `api/form-letters/?assigned_section={assigned_section}&start_date={start_date}&end_date={end_date}` displays information which is consistent with the data on the `/admin` page.
Do the below while NOT logged in as an admin:
[ ] Confirm that you cannot access `/api/form-letters`

### Author

+ [ ] If this is a story, run locally and check to make sure all Acceptance Criteria are met. (If any criteria are unclear, ask about them.).
+ [ ] Check for, document, and establish a testing plan for any behavior that may vary across environments or is otherwise difficult to test.
+ [ ] Check for [accessibility](https://github.com/usdoj-crt/crt-portal/blob/develop/docs/a11y_plan.md).
+ [ ] [Tests pass](https://github.com/USDOJ/crt-portal/#tests).

### Reviewer

+ [ ] If this is a story, run locally and check to make sure all Acceptance Criteria are met. (If any criteria are unclear, ask about them.).
+ [x] Check for any behavior that may vary across environments or is difficult to test, and ensure that it is well-understood, documented, and that there is a testing plan in place.
+ [x] Re-check for [accessibility](https://github.com/usdoj-crt/crt-portal/blob/develop/docs/a11y_plan.md).
+ [x] [Tests pass](https://github.com/USDOJ/crt-portal/#tests).

## Notes for reviewer:

See [PR instructions doc](https://github.com/usdoj/crt-portal/blob/master/docs/pull_requests.md) for full pull request review instructions.

### CODE REVIEW

File: crt_portal/api/filters.py [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+
+
+def contacts_filter(querydict):
+    kwargs = {}
+    filters = {}
+    action_qs = Action.objects.filter().all()
+    contact_qs = Action.objects.filter(verb='Contacted complainant:').all()
+    total_actions = len(action_qs)
+    total_contacts = len(contact_qs)
+    total_emails_counter = {'CRM - R1 Form Letter': 0,
```
Comment (eileen-nava) [inline crt_portal/api/filters.py:23]: Billy and I discussed this while on a call.  Instead of hardcoding the email titles in `total_emails_counter`, I will instead pull the email titles from `/api/responses`.

File: crt_portal/api/tests/test_filters.py [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+from actstream.models import Action
+from django.http import QueryDict
+
+from django.test import TestCase
+from api.tests.test_data import SAMPLE_ACTION_1, SAMPLE_ACTION_2, SAMPLE_ACTION_3
+from ..filters import contacts_filter
+
+
+class ContactsFilterTests(TestCase):
+    def setUp(self):
```
Comment (eileen-nava) [inline crt_portal/api/tests/test_filters.py:11]: Note to self: add a tearDown() method.

File: crt_portal/api/views.py [hunk 3]
```diff
@@ -102,3 +103,15 @@ class ReportCountView(APIView):
     def get(self, request, format=None):
```
Comment (billyhunt) [inline crt_portal/api/views.py:103]: I wonder if we should add your new snazzy url to the api root?

File: crt_portal/api/views.py [hunk 4]
```diff
@@ -102,3 +103,15 @@ class ReportCountView(APIView):
...
         reports_accessed_payload = reports_accessed_filter(request.GET)
         return Response(reports_accessed_payload)
+
+
+class FormLettersIndex(APIView):
+    """
+    A view that displays information about the number of form letters sent.
+    """
+
```
Comment (eileen-nava) [inline crt_portal/api/views.py:112]: TODO: include an example request.  Billy did this in his api work.

File: crt_portal/api/views.py [hunk 5]
```diff
@@ -102,3 +103,15 @@ class ReportCountView(APIView):
...
+    permission_classes = (IsAuthenticated,)
+
+    def get(self, request):
+        contacts_payload = contacts_filter(request.GET)
```
Comment (eileen-nava) [inline crt_portal/api/views.py:116]: todo: add error handling for when a user uses an invalid date format.  Right now, an invalid date format causes all results to be returned.

**Comment by billyhunt** [comment]
It looks as if a poorly formatted date returns all results.  

http://localhost:8000/api/form-letters/?start_date=4-10-2022&end_date=4-20-2022

**Comment by billyhunt** [comment]
AS it stands I think we are returning results with 0.  I wonder if we should prune them.

File: crt_portal/api/views.py [hunk 6]
```diff
@@ -71,6 +72,15 @@ class ResponseList(generics.ListAPIView):
     permission_classes = [permissions.IsAuthenticated]
 
 
+class ResponseTitleList(generics.ListAPIView):
+    """
+    API endpoint that returns response titles.
+    """
+    queryset = ResponseTemplate.objects.all()
+    serializer_class = ResponseTitleSerializer
+    permission_classes = [permissions.IsAuthenticated]
```
Comment (eileen-nava) [inline crt_portal/api/views.py:81]: I thought this line shouldn't be necessary, due to 
```
REST_FRAMEWORK = {
    'DEFAULT_PERMISSION_CLASSES': (
        'rest_framework.permissions.IsAuthenticated',
    )
}
```

However, tests break when I remove `permission_classes = [permissions.IsAuthenticated]`.  I'm not sure what the utility of `DEFAULT_PERMISSION_CLASSES` is - does it accomplish anything?

File: crt_portal/api/filters.py [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+
+    for contact in filtered_contacts:
+        try:
+            email_title = contact.description.split("'")[1]
+            if email_title:
+                for key in emails_counter_for_date_range:
+                    if key == email_title:
+                        emails_counter_for_date_range[key] += 1
+        except IndexError:
+            print("oh my, it is not an email")
```
[the remark this answers is not in the document]
Comment (eileen-nava) [reply crt_portal/api/filters.py:96]: When you're available, I'd like to touch base on what kind of error this should.  I currently am raising a `ParseError`, but I'm not sure if that's the right choice. 🤔

File: crt_portal/api/filters.py [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+
+    contacts_payload['total_contacts_in_range'] = len(filtered_contacts)
+    contacts_payload['total_actions_in_range'] = len(filtered_actions)
+    contacts_payload['total_emails_counter'] = total_emails_counter
+    contacts_payload['emails_counter_for_date_range'] = emails_counter_for_date_range
```
[the remark this answers is not in the document]
Comment (eileen-nava) [reply crt_portal/api/filters.py:101]: I *think* that since I instantiate `emails_counter_for_date_range` on L45, it will exist even if the try fails.  However, maybe I shouldn't create `emails_counter_for_date_range` if there is no date range specified in the path variables?

File: crt_portal/api/tests/test_filters.py [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+        Action.objects.create(**SAMPLE_ACTION_1)
+        Action.objects.create(**SAMPLE_ACTION_2)
+        Action.objects.create(**SAMPLE_ACTION_3)
```
[the remark this answers is not in the document]
Comment (eileen-nava) [reply crt_portal/api/tests/test_filters.py:14]: 🙏🏻

File: crt_portal/cts_forms/filters.py [hunk 10]
```diff
@@ -35,6 +37,10 @@
     'modified_date_start': '__gte',  # not in filter controls?
     'modified_date_end': '__lte',  # not in filter controls?
 
+    # API
+    'start_date': '__gte',
+    'end_date': '__lte',
```
[the remark this answers is not in the document]
Comment (eileen-nava) [reply crt_portal/cts_forms/filters.py:42]: I removed the extra stuff 👍🏻

File: crt_portal/cts_forms/urls.py [hunk 11]
```diff
@@ -23,5 +23,5 @@
     path('attachment/report/<int:report_id>/', ReportAttachmentView.as_view(), name='save-report-attachment'),
     path('attachment/<int:attachment_id>/', RemoveReportAttachmentView.as_view(), name='remove-report-attachment'),
     path('trends/', TrendView.as_view(), name='trends'),
-    path('dashboard/', dashboard_view, name='dashboard'),
+    path('dashboard/', dashboard_view, name='dashboard')
```
Comment (eileen-nava) [inline crt_portal/cts_forms/urls.py:26]: @billyhunt Should I revert this change, too?  I don't want to clash with your work.

**Comment by eileen-nava** [comment]
@billyhunt Can you please review this on Monday?  Thanks!

File: crt_portal/cts_forms/tests/test_filters.py [hunk 12]
```diff
@@ -217,7 +218,8 @@ def test_malformed_parens_search(self):
 
 
 class ReportLanguageFilterTests(TestCase):
-    def setUp(self):
+    @classmethod
+    def setUpTestData(self):
```
Comment (eileen-nava) [inline crt_portal/cts_forms/tests/test_filters.py:222]: @billyhunt Here is an end-of-day status update.  Tests continue to fail due to the duplicate key value violating the primary key restraint on `cts_forms_report_pkey`.  Output below:

```
======================================================================
ERROR: test_reported_language_vi (cts_forms.tests.test_filters.ReportLanguageFilterTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/local/lib/python3.9/site-packages/django/db/backends/utils.py", line 84, in _execute
    return self.cursor.execute(sql, params)
psycopg2.errors.UniqueViolation: duplicate key value violates unique constraint "cts_forms_report_pkey"
DETAIL:  Key (id)=(1) already exists.
```

I will investigate more on Monday.

**Review by billyhunt — CHANGES_REQUESTED** [review]
It looks really good.  Add those API tests, and return 0 results when you do not include an assigned_section, and you are good to go!

File: crt_portal/api/filters.py [hunk 13]
```diff
@@ -1,19 +1,96 @@
...
+
+    form_letters_payload = {
+        "start_date": "",
+        "end_date": "",
+        "assigned_section": "",
+        "total_form_letters": 0,
+        "form_letters_counter": {}
+    }
+
+    section = querydict["assigned_section"]
```
Comment (billyhunt) [inline crt_portal/api/filters.py:54]: This is throwing an exception if you do not have the field on the url.  

image

To recreate, I went to http://localhost:8000/api and clicked the http://localhost:8000/api/form-letters/ link.

**Comment by billyhunt** [comment]
My tests are failing locally on this branch, and I am getting that ```django.core.exceptions.ImproperlyConfigured: The SECRET_KEY setting must not be empty.``` error.  How did you end up resolving that?

**Review by billyhunt — CHANGES_REQUESTED** [review]

File: crt_portal/cts_forms/urls.py (hunk 11 above)
Comment (eileen-nava) [reply crt_portal/cts_forms/urls.py:26]: Thanks! 🙏🏻

File: .circleci/config.yml [hunk 14]
```diff
@@ -403,6 +403,10 @@ jobs:
       - run:
           name: Login to cloud.gov Staging
           command: cf login -u ${CRT_USERNAME_STAGE} -p ${CRT_PASSWORD_STAGE} -o doj-crtportal -s ${CF_SPACE}
+      # Refresh the form letters sent view in staging
+      - run:
+          name: Refresh form letters sent view
+          command: cf run-task crt-portal-django  -c "python crt_portal/manage.py refresh_form_letters_sent" --name refresh-form-letters-sent-view
```
[the remark this answers is not in the document]
Comment (eileen-nava) [reply .circleci/config.yml:409]: Good idea.  I added it to create_mock_reports and run.sh.

File: crt_portal/cts_forms/filters.py [hunk 15]
```diff
@@ -171,44 +162,6 @@ def dashboard_filter(querydict):
     return filters, filtered_actions
 
 
-def reports_accessed_filter(querydict):
```
[the remark this answers is not in the document]
Comment (eileen-nava) [reply crt_portal/cts_forms/filters.py:174]: It is now in `cts_forms/api`, which I like because this filter is called from `api/views.py`.  Let me know if you disagree with the move.

[the remark this answers is not in the document]
Comment (billyhunt) [reply crt_portal/cts_forms/filters.py:174]: No, that makes good horse sense!

**Review by billyhunt — APPROVED** [review]
Looks Rad to Me!
